### PR TITLE
unix: support macOS CoW file copy

### PR DESCRIPTION
**Pull request by juanarbol**

### CODE REVIEW

File: src/unix/fs.c [hunk 1]
```diff
@@ -1241,6 +1242,7 @@ static ssize_t uv__fs_copyfile(uv_fs_t* req) {
   off_t in_offset;
   off_t bytes_written;
   size_t bytes_chunk;
+  char* cow_buf;
```
Comment (saghul) [inline src/unix/fs.c:1245]: This is going to cause a warning on platforms other than apple, maybe declare it inside the if?

Comment (juanarbol) [reply src/unix/fs.c:1245]: I wrote it like that due to the style/lint rules.

Comment (saghul) [reply src/unix/fs.c:1245]: then add a `(void) cow_buf;` in the else ifdef branch pl.

Comment (juanarbol) [reply src/unix/fs.c:1245]: Fixing it!

Comment (juanarbol) [reply src/unix/fs.c:1247]: I think it is not a sin having a conditional for platform-specific variables

https://github.com/libuv/libuv/blob/bf61390769068de603e6deec8e16623efcbe761a/src/unix/stream.c#L404

File: src/unix/fs.c [hunk 2]
```diff
@@ -1374,6 +1379,46 @@ static ssize_t uv__fs_copyfile(uv_fs_t* req) {
...
+        goto out;
+      } else {
+        goto fallback;
+      }
+    }
+
+    /* Write the mapped source file to the destination file */
+    if (write(dstfd, cow_buf, src_statsbuf.st_size) == src_statsbuf.st_size) {
+      /* Everything went fine now unmap the source file */
+      if (munmap(cow_buf, src_statsbuf.st_size) == 0) {
```
[the remark this answers is not in the document]
Comment (juanarbol) [reply src/unix/fs.c:1407]: That is true. That is a good question tho. The clone is done but the map is not released, what should I do in that case?

[the remark this answers is not in the document]
Comment (juanarbol) [reply src/unix/fs.c:1407]: By reading the man pages of `munmap`, it fails under this circumstances.

```
     [EINVAL]           The addr parameter was not page aligned (i.e., a multiple of the page size).

     [EINVAL]           The len parameter was negative or zero.

     [EINVAL]           Some part of the region being unmapped is not part of the currently valid address space.
```

The first two cases are previously validated by the `mmap` call and error handling; and the last one... we have full control over that mmap, I think we can trust the some regions are not being `munmap(ed)` I believe we can trust that `munmap` won't fail.

[the remark this answers is not in the document]
Comment (juanarbol) [reply src/unix/fs.c:1407]: But just to be sure, I guess that If that fails, the better is do `fallback`, even if that is a "re-do but not using CoW" operation. Same as linux does unless `UV_FS_COPYFILE_FICLONE_FORCE` is provided. I believe that the way it is, it covers all the possible scenarios on the best way. But I'm open to modify this if needed.

**Comment by juanarbol** [comment]
Any update here? Should I close this?

**Comment by saghul** [comment]
LGTM anyone else? @libuv/collaborators

**Comment by cjihrig** [comment]
@juanarbol have you tested these changes in Node.js? We used to have a Jenkins job for that, but I'm not sure if it still works.

**Comment by santigimeno** [comment]
Maybe I'm misunderstanding things but is this really equivalent to what `clonefile()` does? If I'm reading this correctly, the `mmap()` call creates a private virtual mapping of the contents of the file (CoW), but then we're writing the data to the destination. Are we sure that after the call to `write()` the data blocks are shared between the src and the dst file?

**Comment by juanarbol** [comment]
> @juanarbol have you tested these changes in Node.js? We used to have a Jenkins job for that, but I'm not sure if it still works.

I can run the job. I'll do that.

**Comment by juanarbol** [comment]
> Maybe I'm misunderstanding things but is this really equivalent to what `clonefile()` does? If I'm reading this correctly, the `mmap()` call creates a private virtual mapping of the contents of the file (CoW), but then we're writing the data to the destination. Are we sure that after the call to `write()` the data blocks are shared between the src and the dst files?

Yeah, that is not a "very strict" implementation, but we could get some performance wins on this one w/out deleting any files or dealing with macOS kernel issues. I just tried to find the "sweet" spot on this implementation, that is easy to implement and safe to use.

**Comment by cjihrig** [comment]
> Once the CI is released again -at this moment there is a security release running-.

It might be faster to just build locally 😄 . Then you could experiment beyond what is currently in the Node test suite. No pressure though if you don't want to spend the time doing that.

**Comment by juanarbol** [comment]
> It might be faster to just build locally 😄 . Then you could experiment beyond what is currently in the Node test suite. No pressure though if you don't want to spend the time doing that.

That is right. HAHAH

At least Node.js main branch seems happy (with [this commit](https://github.com/nodejs/node/commit/66b76e24e2a277b32d65094684ab7410812e45ae) -latest at the moment I comment this). The `make test-only` script is happy.

**Comment by bnoordhuis** [comment]
I don't think this approach can work for a library like libuv. The problem with mmap is that you get a SIGBUS or similar signal if the file is truncated concurrently by another process.

**Comment by juanarbol** [comment]
I had no idea about that mmap behaviour, thanks! I'll close this then.

**Pull request closed without merging**